### PR TITLE
Adding an example to demonstrate the CDS Library API in C/C++11

### PR DESCRIPTION
**Pull request by pralhadsapre**
### Summary
New example directory for `cloud_discovery_service` showing example usage of the Library API in C/C++11.

### Details and comments
Sample code to show how CDS can be integrated by users in their applications using C or C++11.

### Checks
-   [x] I have updated the documentation accordingly.
-   [x] I have read the [CONTRIBUTING](https://github.com/rticommunity/rticonnextdds-examples/blob/develop/CONTRIBUTING.md) document.

### CODE REVIEW

File: examples/cloud_discovery_service/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# (c) 2023 Copyright, Real-Time Innovations, Inc. All rights reserved.
+# No duplications, whole or partial, manual or electronic, may be made
+# without express written permission.  Any such copies, or revisions thereof,
+# must display this notice unaltered.
+# This code contains trade secrets of Real-Time Innovations, Inc.
+
+cmake_minimum_required(VERSION 3.11)
```
Comment (lulivi) [inline examples/cloud_discovery_service/CMakeLists.txt:7]:
```suggestion
cmake_minimum_required(VERSION 3.12)
```

File: examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,60 @@
+# (c) 2023 Copyright, Real-Time Innovations, Inc. All rights reserved.
+# No duplications, whole or partial, manual or electronic, may be made
+# without express written permission. Any such copies, or revisions thereof,
+# must display this notice unaltered.
+# This code contains trade secrets of Real-Time Innovations, Inc.
+
+cmake_minimum_required(VERSION 3.11)
```
Comment (lulivi) [inline examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt:7]:
```suggestion
cmake_minimum_required(VERSION 3.12)
```

File: examples/cloud_discovery_service/CMakeLists.txt (hunk 1 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/CMakeLists.txt:7]: I see. Done.

File: examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt (hunk 2 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt:7]: Done.

File: examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+target_link_directories(CloudDiscoveryServiceLibraryAPICXX11
+    PUBLIC
+        ${CONNEXTDDS_DIR}/lib/${CONNEXTDDS_ARCH}
+)
+
+target_link_libraries(CloudDiscoveryServiceLibraryAPICXX11
+    PUBLIC
+        RTIConnextDDS::cpp2_api
+        rticlouddiscoveryservice
+)
```
Comment (lulivi) [inline examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt:60]: Apply once https://github.com/rticommunity/rticonnextdds-cmake-utils/pull/76 is merged.

```suggestion
        cloud_discovery_service
)

add_executable(CloudDiscoveryServiceLibraryAPICXX11
    "${CMAKE_CURRENT_SOURCE_DIR}/cloud_discovery_service_library_api.cxx"
)

set_property(
    TARGET CloudDiscoveryServiceLibraryAPICXX11
    PROPERTY CXX_STANDARD 11
)
set_property(
    TARGET CloudDiscoveryServiceLibraryAPICXX11
    PROPERTY CXX_STANDARD_REQUIRED ON
)

target_link_libraries(CloudDiscoveryServiceLibraryAPICXX11
    PUBLIC
        RTIConnextDDS::cloud_discovery_service_cpp2
)

```

**Comment by lulivi** [comment]
Blocked by https://github.com/rticommunity/rticonnextdds-examples/pull/578

**Comment by lulivi** [comment]
Hi @pralhadsapre can you add Cloud Discovery Service directory to the main `examples/CMakeLists.txt` file as it is done in [this](https://github.com/rticommunity/rticonnextdds-examples/pull/575/files#diff-940e5356fa1137e646ea135a6f68cacbfad4fe7124c3a69163468f588acf9283) PR?

File: examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt (hunk 3 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/CMakeLists.txt]: Done.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    // To turn on additional logging, include <rti/config/Logger.hpp> and
+    // uncomment the following line:
+    // rti::config::Logger::instance().verbosity(rti::config::Verbosity::STATUS_ALL);
+
+    try {
+        service_main(cfg_name, cfg_file);
+    } catch (const std::exception &ex) {
+        // This will catch DDS exceptions
+        std::cerr << "Exception in service_main: " << ex.what() << std::endl;
+        return -1;
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:80]:
```suggestion
        return EXIT_FAILURE;
```

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    return 0;
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:83]:
```suggestion
    return EXIT_SUCCESS;
```

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ of any type, including any warranty for fitness for any purpose. RTI is under
+ no obligation to maintain or support the Software. RTI shall not be liable for
+ any incidental or consequential damages arising out of the use or inability to
+ use the software.
+ ******************************************************************************/
+
+#include <iostream>
+#include <thread>
+#include <chrono>
+#include <map>
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:15]:
```suggestion
#include <map>
#include <cstdlib>
```

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+#include <rti/cds/Service.hpp>
+
+void service_main(std::string cfg_name, std::string cfg_file)
+{
+    rti::cds::Service service(
+            rti::cds::ServiceProperty()
+                    .cfg_file(cfg_file)
+                    .service_name(cfg_name)
+                    .application_name(
+                            "CloudDiscoveryService-LibraryAPI-CXX11"));
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:26]: Perhaps it'd be more clear to separate this in two statements: one declaring the ServiceProperty and one to create the Service.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                    rti::config::Verbosity::STATUS_ALL));
+
+    service.start();
+
+    std::cout << "ConfigName=" << cfg_name << ", "
+              << "ConfigFile=" << cfg_file << std::endl;
+    // Infinite loop until the user terminates the process
+    while (1) {
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+    }
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:40]: We should have an exit condition  (handling CTRL+C) and maybe a timeout option, but the latter wouldn't be needed.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+int main(int argc, char *argv[])
+{
+    const std::string DEFAULT_CFG_NAME = "LibraryAPIDemo";
+    const std::string DEFAULT_CFG_FILE = "CloudDiscoveryServiceConfig.xml";
+
+    std::string cfg_name = DEFAULT_CFG_NAME;
+    std::string cfg_file = DEFAULT_CFG_FILE;
+
+    for (int i = 1; i < argc;) {
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:53]: Seems simple enough but you could have some separate function to parse command-line arguments. Not necessary probably.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 10]
```diff
@@ -0,0 +1,84 @@
+/*******************************************************************************
+ (c) 2005-2023 Copyright, Real-Time Innovations, Inc. All rights reserved.
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:2]:
```suggestion
 (c) 2023 Copyright, Real-Time Innovations, Inc. All rights reserved.
```

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        NDDS_Utility_sleep(&sleepPeriod);
+    }
+
+    ok = RTI_CDS_Service_stop(service);
+    if (!ok) {
+        printf("Error stopping the Cloud Discovery Service instance\n");
+        return -1;
+    }
+
+    /* Sample usage of deleting the service instance for resource cleanup */
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:59]: We could have a cleanup function or a goto done here.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+    /* Dial up the logging at service level */
+    property.service_verbosity = RTI_CDS_LOG_VERBOSITY_ALL;
+    property.dds_verbosity = RTI_CDS_LOG_VERBOSITY_EXCEPTIONS;
+
+
+    service = RTI_CDS_Service_new(&property);
+    if (service == NULL) {
+        printf("Error creating the Cloud Discovery Service instance\n");
+        return -1;
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:38]: Use `EXIT_FAILURE`

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+    ok = RTI_CDS_Service_start(service);
+    if (!ok) {
+        printf("Error starting the Cloud Discovery Service instance\n");
+        return -1;
+    }
+
+    printf("ConfigName=%s, ConfigFile=%s\n", cfgName, cfgFile);
+    /* Infinite loop until the user terminates the process */
+    while (1) {
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:49]: Handle CTRL+C?

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    /* RTI_CDS_Service_delete(service); */
+
+    /* Sample usage of deleting the service property to free the strings */
+    /* RTI_CDS_Property_finalize(&property); */
+}
+
+int main(int argc, char *argv[])
+{
+#define DEFAULT_CFG_NAME "LibraryAPIDemo"
+#define DEFAULT_CFG_FILE "CloudDiscoveryServiceConfig.xml"
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:69]: Could we simply use the constant?

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    if (!ok) {
+        printf("Error stopping the Cloud Discovery Service instance\n");
+        return -1;
+    }
+
+    /* Sample usage of deleting the service instance for resource cleanup */
+    /* RTI_CDS_Service_delete(service); */
+
+    /* Sample usage of deleting the service property to free the strings */
+    /* RTI_CDS_Property_finalize(&property); */
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:63]: We're missing a return,
```suggestion
    /* RTI_CDS_Property_finalize(&property); */
    
    return EXIT_SUCCESS;
```

**Review by fgaranda — CHANGES_REQUESTED** [review]

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    // rti::config::Logger::instance().verbosity(rti::config::Verbosity::STATUS_ALL);
+
+    try {
+        service_main(cfg_name, cfg_file);
+    } catch (const std::exception &ex) {
+        // This will catch DDS exceptions
+        std::cerr << "Exception in service_main: " << ex.what() << std::endl;
+        return -1;
+    }
+
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:82]: Should we finalize the participant factory before exiting?

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 4 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:80]: Done.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 5 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:83]: Done.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 7 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:26]: Sure. Done.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 12 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:38]: Done.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 16 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:82]: Done.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 14 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:69]: I need those values for default initialization of the variables and in the help text.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 10 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:2]: Done.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 15 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:63]: Good catch. Done.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 6 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:15]: We need this for the ServiceProperty field `std::map<std::string, std::string> user_environment`.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 13 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:49]: I added a `runForSecs` parameter as we discussed on Slack.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 11 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:59]: I added a `runForSecs` parameter as we discussed on Slack. Now the cleanup code is executed all the time.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 9 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:53]: True. We should in theory migrate the command line parsing to use the `app_utils.1.0` APIs, but I guess for now this works fine.

File: examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx (hunk 8 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/cloud_discovery_service_library_api.cxx:40]: I added a `runForSecs` parameter as we discussed on Slack.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+    /* Sleep for the running period */
+    NDDS_Utility_sleep(&runningPeriod);
+
+    ok = RTI_CDS_Service_stop(service);
+    if (!ok) {
+        printf("Error stopping the Cloud Discovery Service instance\n");
+        return EXIT_FAILURE;
+    }
+
+    /* Deleting the service instance for resource cleanup */
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:63]: You won't get here if some of the error conditions return. You can either use a `goto done` or create a cleanup function.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    RTI_CDS_Property_finalize(&property);
+
+    /*
+     * RTI Data Distribution Service provides the finalize_instance() method on
+     * domain participant factory for users who want to release memory used
+     * by the participant factory. Uncomment the following block of code for
+     * clean destruction of the singleton.
+     */
+    /*
+    DDS_ReturnCode_t retcode;
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:76]: Perhaps we should move this outside the scope of this function, like we did in the previous example. That shows how you'd do it in a regular application, so you don't affect other services or native DomainParticipants.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 14 above)
Comment (fgaranda) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c:90]: Shouldn't those be constants and the cfgFile and cfgName pointers pointing to either the default or argv[i++]?

**Review by fgaranda — CHANGES_REQUESTED** [review]

File: examples/cloud_discovery_service/library_api/c++11/CloudDiscoveryServiceConfig.xml [hunk 19]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<dds xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+      xsi:noNamespaceSchemaLocation="https://community.rti.com/schema/current/rti_cloud_discovery_service.xsd">
```
Comment (fgaranda) [inline examples/cloud_discovery_service/library_api/c++11/CloudDiscoveryServiceConfig.xml:3]:
```suggestion
      xsi:noNamespaceSchemaLocation="https://community.rti.com/schema/7.0.0/rti_cloud_discovery_service.xsd">
```
We do a bulk change later on

Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/CloudDiscoveryServiceConfig.xml:3]: Do we typically do a bulk change of all the XSD versions with each release?

Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c++11/CloudDiscoveryServiceConfig.xml:3]: Alright .. I will make the change.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 18 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c]: I moved this to a `service_shutdown()` function.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 17 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c]: I moved this to a `service_shutdown()` function.

File: examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c (hunk 14 above)
Comment (pralhadsapre) [reply examples/cloud_discovery_service/library_api/c/cloud_discovery_service_library_api.c]: I see .. I removed the `#define`s and directly assigned the variables these constants.

**Review by fgaranda — DISMISSED** [review]

**Review by fgaranda — APPROVED** [review]

**Review by lulivi — APPROVED** [review]
- [x] Wait for the validation

**Review by fercsrti77 — DISMISSED** [review]

**Review by lulivi — APPROVED** [review]

**Pull request merged**